src/net/direct.c

```diff
@@ -118,12 +120,12 @@ closure_function(1, 1, status, direct_conn_send,
         u64 flags = irq_disable_save();
         list_insert_before(&dc->sendq_head, &q->l); /* really need CAS version */
         irq_restore(flags);
-        if (!direct_conn_send_internal(dc)) {
-            if (!dc->running) {
-                thunk t = closure(dc->d->h, direct_conn_send_bh, dc);
-                dc->running = true;
-                assert(enqueue(runqueue, t));
-            }
+        if (!dc->send_timer && !direct_conn_send_internal(dc)) {
+            timer_handler th = closure(dc->d->h, direct_conn_send_timer, dc);
+            assert(th != INVALID_ADDRESS);
+            dc->send_timer = register_timer(runloop_timers, CLOCK_ID_MONOTONIC, DIRECT_CONN_SEND_INTERVAL,
+                                            false, true, th);
+            assert(dc->send_timer != INVALID_ADDRESS);
```

Comment:
Is there a reason we can't just use lwip's tcp_sent on the direct connection to specify a callback when send buffer frees up instead of using the timer and retry the send there?

Reply:
You're absolutely right. Going to close this and re-open another PR...